### PR TITLE
[ApiManagement] Secure Parameter for Cmdlets

### PR DESCRIPTION
**Pull request by solankisamir**
## Description
* Breaking Changes in Cmdlet to Manage Api Management Users https://github.com/Azure/azure-powershell/issues/4595
    - New-AzureRmApiManagementUser Parameter `Password` is changed from String to SecureString
    - Set-AzureRmApiManagementBackend Parameter `Password` is changed from String to SecureString
    
* Breaking Changes in Cmdlet to Create Backend Proxy Object
    - New-AzureRmApiManagementBackendProxy Parameter `Password` and `UserName` has been replaced with `ProxyCredentials` of type PSCredential
	
* Updated Cmdlet Get-AzureRmApiManagementUser to fix issue https://github.com/Azure/azure-powershell/issues/4510

* Updated Cmdlet New-AzureRmApiManagementApi to create Api with Empty Path https://github.com/Azure/azure-powershell/issues/4069

* Updated documentation for issue https://github.com/Azure/azure-powershell/issues/4676

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [ ] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

File: src/ResourceManager/ApiManagement/Commands.ApiManagement/help/Get-AzureRmApiManagementUser.md [hunk 1]
```diff
@@ -35,7 +38,7 @@ The **Get-AzureRmApiManagementUser** cmdlet gets a specified user, or all users,
 
 ### Example 1: Get all users
 ```
-PS C:\>Get-AzureRmApiManagementUser -Context $apimContext
+PS C:\>Get-AzureRmApiManagementUser -Context $apimContext
```
[the remark this answers is not in the document]
Comment (solankisamir) [reply src/ResourceManager/ApiManagement/Commands.ApiManagement/help/Get-AzureRmApiManagementUser.md]: They are already there.

File: src/ResourceManager/ApiManagement/Commands.ApiManagement/help/New-AzureRmApiManagementBackend.md [hunk 2]
```diff
@@ -26,6 +26,7 @@ Creates a new backend entity in Api Management.
 
 ### Create Backend 123 with a Basic Authorization Scheme
 ```
+PS C:\>$apimContext = New-AzureRmApiManagementContext -ResourceGroupName "Api-Default-WestUS" -ServiceName "contoso"
 $credential = New-AzureRmApiManagementBackendCredential -AuthorizationHeaderScheme basic -AuthorizationHeaderParameter opensesame -Query @{"sv" = @('xx', 'bb'); "sr" = @('cc')} -Header @{"x-my-1" = @('val1', 'val2')}
```
[the remark this answers is not in the document]
Comment (solankisamir) [reply src/ResourceManager/ApiManagement/Commands.ApiManagement/help/New-AzureRmApiManagementBackend.md]: done

File: src/ResourceManager/ApiManagement/Commands.ApiManagement/help/New-AzureRmApiManagementBackendProxy.md [hunk 3]
```diff
@@ -23,21 +23,22 @@ Creates a new Backend Proxy Object which can be piped when creating a new Backen
 
 ### Create a Backend Proxy In-Memory Object
 ```
-$proxy= New-AzureRmApiManagementBackendProxy -Url "https://abbc.def.g" -UserName "apim" -Password "password"
```
[the remark this answers is not in the document]
Comment (solankisamir) [reply src/ResourceManager/ApiManagement/Commands.ApiManagement/help/New-AzureRmApiManagementBackendProxy.md]: done

File: src/ResourceManager/ApiManagement/Commands.ApiManagement/help/New-AzureRmApiManagementContext.md [hunk 4]
```diff
@@ -24,13 +25,28 @@ The context is used for all of the API Management service cmdlets.
 
 ### Example 1: Create a PsApiManagementContext instance
 ```
-PS C:\>$ApiMgmtContext = New-AzureRmApiManagementContext -ResourceGroupName "ContosoResources" -ServiceName "Contoso"
+PS C:\>PS C:\>$ApiMgmtContext = New-AzureRmApiManagementContext -ResourceGroupName "ContosoResources" -ServiceName "Contoso"
```
[the remark this answers is not in the document]
Comment (solankisamir) [reply src/ResourceManager/ApiManagement/Commands.ApiManagement/help/New-AzureRmApiManagementContext.md]: done

File: src/ResourceManager/ApiManagement/Commands.ApiManagement/help/New-AzureRmApiManagementUser.md [hunk 5]
```diff
@@ -25,7 +25,8 @@ The **New-AzureRmApiManagementUser** cmdlet registers a new user.
 
 ### Example 1: Register a new user
 ```
-PS C:\>New-AzureRmApiManagementUser -Context $apimContext -FirstName "Patti" -LastName "Fuller" -Email "Patti.Fuller@contoso.com" -Password "qwerty"
+$securePassword = ConvertTo-SecureString "qwerty" -AsPlainText -Force
```
[the remark this answers is not in the document]
Comment (solankisamir) [reply src/ResourceManager/ApiManagement/Commands.ApiManagement/help/New-AzureRmApiManagementUser.md]: done

File: src/ResourceManager/ApiManagement/Commands.ApiManagement/help/Remove-AzureRmApiManagementIdentityProvider.md [hunk 6]
```diff
@@ -22,8 +23,8 @@ Removes an existing Identity Provider Configuration.
 ## EXAMPLES
 
 ### Removes the Facebook identity provider settings from ApiManagement service
-
 ```
+PS C:\>$apimContext = New-AzureRmApiManagementContext -ResourceGroupName "Api-Default-WestUS" -ServiceName "contoso"
 Remove-AzureRmApiManagementIdentityProvider -Context $apimContext -Type 'Facebook' -PassThru
```
[the remark this answers is not in the document]
Comment (solankisamir) [reply src/ResourceManager/ApiManagement/Commands.ApiManagement/help/Remove-AzureRmApiManagementIdentityProvider.md]: done

File: src/ResourceManager/ApiManagement/Commands.ApiManagement/help/Set-AzureRmApiManagementUser.md [hunk 7]
```diff
@@ -25,7 +25,9 @@ The **Set-AzureRmApiManagementUser** cmdlet sets user details.
 
 ### Example 1: Change a user's password, email address and state
 ```
-PS C:\>Set-AzureRmApiManagementUser -Context $apimContext -UserId "0123456789" -Email "patti.fuller@contoso.com" -Password "asdfgh" -State "Blocked"
+PS C:\>$apimContext = New-AzureRmApiManagementContext -ResourceGroupName "Api-Default-WestUS" -ServiceName "contoso"
+$securePassword = ConvertTo-SecureString "qwerty" -AsPlainText -Force
```
[the remark this answers is not in the document]
Comment (solankisamir) [reply src/ResourceManager/ApiManagement/Commands.ApiManagement/help/Set-AzureRmApiManagementUser.md]: done

File: src/ResourceManager/ApiManagement/ChangeLog.md [hunk 8]
```diff
@@ -18,6 +18,16 @@
         - Additional information about change #1
 -->
 ## Current Release
+* Breaking Changes in Cmdlet to Manage Api Management Users
+    - New-AzureRmApiManagementUser Parameter `Password` is changed from String to SecureString
+    - Set-AzureRmApiManagementBackend Parameter `Password` is changed from String to SecureString
+    
+* Breaking Changes in Cmdlet to Create Backend Proxy Object
+    - New-AzureRmApiManagementBackendProxy Parameter `Password` and `UserName` has been replaced with `ProxyCredentials` of type PSCredential
```
[the remark this answers is not in the document]
Comment (solankisamir) [reply src/ResourceManager/ApiManagement/ChangeLog.md]: done

File: src/ResourceManager/ApiManagement/Commands.ApiManagement/help/Get-AzureRmApiManagementBackend.md [hunk 9]
```diff
@@ -13,40 +13,34 @@ Get the details of the Backend.
...
-
-
-
-
+### Example 1: Get all Backends
 
 ```
-Get-AzureRmApiManagementBackend -Context $apimContext
+PS C:\>$apimContext = New-AzureRmApiManagementContext -ResourceGroupName "Api-Default-WestUS" -ServiceName "contoso"
+Get-AzureRmApiManagementBackend -Context $apimContext
```
[the remark this answers is not in the document]
Comment (solankisamir) [reply src/ResourceManager/ApiManagement/Commands.ApiManagement/help/Get-AzureRmApiManagementBackend.md]: done

File: src/ResourceManager/ApiManagement/Commands.ApiManagement/help/Get-AzureRmApiManagementIdentityProvider.md [hunk 10]
```diff
@@ -13,40 +13,30 @@ Get the identity provider configuration details.
...
-### --------------------------  Example 1  --------------------------
-@{paragraph=PS C:\\\>}
-
-
-
-
+### Example 1: Get all Identity Providers
 
 ```
 Get-AzureRmApiManagementIdentityProvider -Context $context
```
[the remark this answers is not in the document]
Comment (solankisamir) [reply src/ResourceManager/ApiManagement/Commands.ApiManagement/help/Get-AzureRmApiManagementIdentityProvider.md]: done

File: src/ResourceManager/ApiManagement/Commands.ApiManagement/help/Get-AzureRmApiManagementTenantAccess.md [hunk 11]
```diff
@@ -23,7 +24,7 @@ The **Get-AzureRmApiManagementTenantAccess** cmdlet gets the tenant access confi
 
 ### Example 1: Get tenant access configuration
 ```
-PS C:\>Get-AzureRmApiManagementTenantAccess -Context $ApimContext
+PS C:\>Get-AzureRmApiManagementTenantAccess -Context $apimContext
```
[the remark this answers is not in the document]
Comment (solankisamir) [reply src/ResourceManager/ApiManagement/Commands.ApiManagement/help/Get-AzureRmApiManagementTenantAccess.md]: added wherever it was missing.

File: src/ResourceManager/ApiManagement/Commands.ApiManagement/help/Get-AzureRmApiManagementUser.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (maddieclayton) [reply src/ResourceManager/ApiManagement/Commands.ApiManagement/help/Get-AzureRmApiManagementUser.md]: Awesome thanks.

File: src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands/GetAzureApiManagementUser.cs [hunk 12]
```diff
@@ -89,6 +89,11 @@ public override void ExecuteApiManagementCmdlet()
                 var user = Client.UserById(Context, UserId);
                 WriteObject(user);
             }
+            else if (ParameterSetName.Equals(FindBy))
+            {
+                var user = Client.UsersList(Context, FirstName, LastName, Email, State, GroupId);
+                WriteObject(user, true);
+            }
             else
```
[the remark this answers is not in the document]
Comment (solankisamir) [reply src/ResourceManager/ApiManagement/Commands.ApiManagement.ServiceManagement/Commands/GetAzureApiManagementUser.cs]: done

File: src/ResourceManager/ApiManagement/Commands.ApiManagement/help/Get-AzureRmApiManagementBackend.md [hunk 13]
```diff
@@ -13,40 +13,34 @@ Get the details of the Backend.
...
-@{paragraph=PS C:\\\>}
-
-
-
-
+### Example 2: Get the Backend specified by the Identifier 123
 
 ```
+PS C:\>$apimContext = New-AzureRmApiManagementContext -ResourceGroupName "Api-Default-WestUS" -ServiceName "contoso"
 Get-AzureRmApiManagementBackend -Context $apimContext -backendId 123
```
[the remark this answers is not in the document]
Comment (solankisamir) [reply src/ResourceManager/ApiManagement/Commands.ApiManagement/help/Get-AzureRmApiManagementBackend.md]: done

File: src/ResourceManager/ApiManagement/Commands.ApiManagement/help/Get-AzureRmApiManagementIdentityProvider.md [hunk 14]
```diff
@@ -13,40 +13,30 @@ Get the identity provider configuration details.
...
-### --------------------------  Example 2  --------------------------
-@{paragraph=PS C:\\\>}
-
-
-
-
-
+### Get the AAD Type Identity Provider
 ```
 Get-AzureRmApiManagementIdentityProvider -Context $context -Type Aad
```
[the remark this answers is not in the document]
Comment (solankisamir) [reply src/ResourceManager/ApiManagement/Commands.ApiManagement/help/Get-AzureRmApiManagementIdentityProvider.md]: done

File: src/ResourceManager/ApiManagement/Commands.SMAPI.Test/ScenarioTests/ApiManagementTests.ps1 [hunk 15]
```diff
@@ -2039,10 +2043,9 @@ Param($resourceGroupName, $serviceName)
     $backendId = getAssetName
     try
     {
```
[the remark this answers is not in the document]
Comment (solankisamir) [reply src/ResourceManager/ApiManagement/Commands.SMAPI.Test/ScenarioTests/ApiManagementTests.ps1]: fixed formatting throughout the file using VSCode extension for Powershell

**Comment by maddieclayton** [comment]
On-demand run here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/25/

**Comment by maddieclayton** [comment]
On demand run here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/28/

**Comment by KedarJoshi** [comment]
#sign-off

**Pull request merged**